### PR TITLE
Include <button type=submit> data whilst constructing the form dataset

### PR DESCRIPTION
**Pull request by Manishearth**
This makes it possible to close things in github (see
https://github.com/Manishearth/mitochondria/issues/1)
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

Either:
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11333)

### CODE REVIEW

File: components/script/dom/htmlbuttonelement.rs [hunk 1]
```diff
@@ -137,6 +138,41 @@ impl HTMLButtonElementMethods for HTMLButtonElement {
...
+        }
+        // Step 3.2
+        let ty = self.Type();
+        // Step 3.4
+        let name = self.Name();
+
+        if name == "" {
+            // unspecced, however all the browsers do it
+            // and it could break urlencoded parsing if
+            // there are random equal signs floating around
```
Comment (Manishearth) [inline components/script/dom/htmlbuttonelement.rs]: I will file a spec bug on this in a while

**Comment by nox** [comment]
Thanks for your contribution, just some nits to address.

-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/dom/htmlbuttonelement.rs, line 145 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11333#-KIRw5X2uzk4-GbiXyTd:-KIRw5X2uzk4-GbiXyTe:131843320) ([raw file](https://github.com/servo/servo/blob/15703f274cfa27c69d4cf57ad22dd33b91c86c86/components/script/dom/htmlbuttonelement.rs#L145)):*
> ```Rust
>     /// Steps range from 3.1 to 3.7 (specific to HTMLButtonElement)
>     pub fn form_datum(&self, submitter: Option<FormSubmitter>) -> Option<FormDatum> {
>         // 3.1: disabled state check is in get_unclean_dataset
> ```

Nit: "Step 3.1"

---

*[components/script/dom/htmlbuttonelement.rs, line 147 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11333#-KIRw7J0vWxl4dkgHBHf:-KIRw7J1k_xIvPD8zVxU:131843320) ([raw file](https://github.com/servo/servo/blob/15703f274cfa27c69d4cf57ad22dd33b91c86c86/components/script/dom/htmlbuttonelement.rs#L147)):*
> ```Rust
>         // 3.1: disabled state check is in get_unclean_dataset
> 
>         // 3.1: only run steps if this is the submitter
> ```

Nit: "Step 3.1"

---

*[components/script/dom/htmlbuttonelement.rs, line 160 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11333#-KIRvQjzGSziHRPT3UDk:-KIRvQjzGSziHRPT3UDl:697145124) ([raw file](https://github.com/servo/servo/blob/15703f274cfa27c69d4cf57ad22dd33b91c86c86/components/script/dom/htmlbuttonelement.rs#L160)):*
> ```Rust
>         let name = self.Name();
> 
>         if name == "" {
> ```

Nit: `name.is_empty()`.

---

*[components/script/dom/htmlbuttonelement.rs, line 167 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11333#-KIRw2CMIoVRt5LgR63i:-KIRw2CMIoVRt5LgR63j:778449479) ([raw file](https://github.com/servo/servo/blob/15703f274cfa27c69d4cf57ad22dd33b91c86c86/components/script/dom/htmlbuttonelement.rs#L167)):*
> ```Rust
>         }
> 
>         // Step 3.6
> ```

AFAICT, this is step 3.9. not 3.6..

---

*[components/script/dom/htmlbuttonelement.rs, line 161 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11333#-KIRwWprVx-iTen0n9-x:-KIRwWprVx-iTen0n9-y:131843320) ([raw file](https://github.com/servo/servo/blob/2cd95645aca276fd4d360e8c1c11a4f81356b5f5/components/script/dom/htmlbuttonelement.rs#L161)):*
> ```Rust
> 
>         if name == "" {
>             // 3.1: Must have a name
> ```

Nit: "Step 3.1"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11333#-:-KIRwYlS5GsJCO9MQfpn:723845725)*

**Comment by nox** [comment]
@bors-servo r+

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11333#-:-KIS2jOMdZKN2yHTAeQs:1236647940)*

**Pull request merged**